api/samples/bbv.cpp

```diff
@@ -0,0 +1,342 @@
+            bb_id_count_pair_t *pair = (bb_id_count_pair_t *)drvector_get_entry(bbv, j);
+
+            const char *format_string;
+            // Only one BB and related hit count for this BBV.
+            if (j == 0 && j == bbv->entries - 1) {
+                format_string = one_pair_only;
+            } else if (j == 0) {
+                format_string = first_pair;
+            } else if (j == bbv->entries - 1) {
+                format_string = last_pair;
+            } else {
+                format_string = middle_pair;
+            }
+
+            char msg[64];
+            int len = dr_snprintf(msg, sizeof(msg) / sizeof(msg[0]), format_string,
+                                  pair->id, pair->count);
+            DR_ASSERT(len > 0);
+
+            dr_write_file(bbvs_file, msg, (size_t)len);
+        }
+    }
+
+    dr_close_file(bbvs_file);
+
+    // Free DR memory.
+    hashtable_delete(&pc_to_id_map);
+    hashtable_delete(&bb_size_table);
+    hashtable_delete(&hit_count_table);
+    if (!drvector_delete(&bbvs))
+        FATAL("ERROR: BBVs drvector not deleted");
+
+    drx_exit();
+    drmgr_exit();
+
+    // Avoid accumulation of option values on static-link re-attach.
+    droption::droption_parser_t::clear_values();
+}
+
+static dr_emit_flags_t
+event_app_instruction(void *drcontext, void *tag, instrlist_t *bb, instr_t *inst,
+                      bool for_trace, bool translating, void *user_data)
+{
+    // By default drmgr enables auto-predication, which predicates all instructions
+    // with the predicate of the current instruction on ARM. We disable it here
+    // because we want to unconditionally execute the following instrumentation.
+    drmgr_disable_auto_predication(drcontext, bb);
+    if (!drmgr_is_first_instr(drcontext, inst))
+        return DR_EMIT_DEFAULT;
+
+    // TODO i#7685: don't rely on absolute PC values. Use drmodtrack library to compute
+    // relative offset instead.
+
+    // Get the BB ID.
+    app_pc bb_pc = instr_get_app_pc(instrlist_first_app(bb));
+    void *bb_id_ptr = hashtable_lookup(&pc_to_id_map, (void *)bb_pc);
+    int bb_id = (int)(ptr_int_t)bb_id_ptr;
+    if (bb_id_ptr == NULL) {
+        bb_id = unique_bb_count;
+        hashtable_add(&pc_to_id_map, (void *)bb_pc, (void *)(ptr_int_t)bb_id);
+        ++unique_bb_count;
+    }
+
+    // TODO i#7685: use hit_count_ptr when we inline the counter increments.
+    int *hit_count_ptr =
+        (int *)hashtable_lookup(&hit_count_table, (void *)(ptr_int_t)bb_id);
+    if (hit_count_ptr == NULL) {
+        // If hit count is not mapped to this BB, then add a new count to the table.
+        int *hit_count = (int *)dr_global_alloc(sizeof(int));
+        *hit_count = 0;
+        hashtable_add(&hit_count_table, (void *)(ptr_int_t)bb_id, hit_count);
+        hit_count_ptr = hit_count;
+    }
+
+    // Get the number of instructions in the BB.
+    int *bb_size_ptr = (int *)hashtable_lookup(&bb_size_table, (void *)(ptr_int_t)bb_id);
```

Comment:
This is technically unsafe as the same start PC could have a different sized block after it, even w/o a JIT: app could load lib A then unload and load lib B in overlapping place; or even re-load A but run code in a different order; or client could flush all the app code and then it runs in different order.

Better: pass the bb size as the user_data (from analysis event) and get rid of this hashtable altogether: both more efficient and more correct in corner cases.

Once this is inlined the size will also be known as a literal right there.